### PR TITLE
Prevent automatic cuda init in init_rpc

### LINKED ISSUE #80141: rpc_init always enables CUDA if it's available
### 🐛 Describe the bug

When initializing RPC, it appears that CUDA is automatically initialized even if an empty device list (i.e., not `None`) is specified in the backend options.  While `None` implies default behavior, an empty list should conceptually mean that no CUDA devices are to be used by RPC, and therefore the RPC backend should not automatically init CUDA.  Setting the environment variable `CUDA_VISIBLE_DEVICES=""` avoids this particular problem, but then prevents the application from initializing/using CUDA for other operations not managed by RPC.

```python
import os
import torch
from torch.distributed import rpc
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
opts = rpc.TensorPipeRpcBackendOptions(devices=[]) # empty list, not None
torch.cuda.is_initialized() # returns False
rpc.init_rpc('worker0', world_size=1, rank=0, rpc_backend_options=opts)
torch.cuda.is_initialized() # returns True
```

I'm running on a Jetson AGX Xavier device, with Ubuntu 18.04, Python 3.8.0 in a virtualenv, with a custom build of PyTorch v1.10.2.  The script below reports clang 6, but in fact PyTorch is compiled with clang 8.

Thanks!

### Versions

Collecting environment information...
PyTorch version: 1.10.0a0+git71f889c
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (aarch64)
GCC version: (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.0 (default, Dec  9 2021, 17:53:27)  [GCC 8.4.0] (64-bit runtime)
Python platform: Linux-4.9.253-tegra-aarch64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/aarch64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.10.0a0+git71f889c
[pip3] torchvision==0.12.0
[conda] Could not collect

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @jjlilley @mrzzd

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80180

Fixes #80141

Only initialize cuda if there are devices specified in `init_rpc`

Differential Revision: [D37458309](https://our.internmc.facebook.com/intern/diff/D37458309)

### CODE REVIEW

File: torch/distributed/rpc/backend_registry.py [hunk 1]
```diff
@@ -262,6 +262,8 @@ def _get_device_infos():
     agent = cast(TensorPipeAgent, api._get_current_rpc_agent())
     opts = agent._get_backend_options()
     device_count = torch.cuda.device_count()
+    if torch.cuda.is_available() and opts.devices:
+        torch.cuda.init()
```
Comment (rohan-varma) [inline torch/distributed/rpc/backend_registry.py:266]: Is there a reason we add torch.cuda.init() here in addition to the call in the tensorpipe backend handler? Could we possibly have it in just one place?

**Review by rohan-varma — COMMENTED** [review]
LGTM, just one question for your consideration

File: torch/distributed/rpc/backend_registry.py (hunk 1 above)
Comment (H-Huang) [reply torch/distributed/rpc/backend_registry.py:266]: One is for statically sized groups, the other is for dynamic/elastic groups (devices are updated as workers join, therefore we need to do another cuda init if new devices are added) so they have different initialization paths and we can't put it in the same place for now.

**Review by rohan-varma — APPROVED** [review]
thanks for the fix!

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37458309).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37458309).

**Pull request closed without merging**